src/trace/implementations/merge_batcher.rs

```diff
@@ -179,81 +143,235 @@ impl<D: Ord, T: Ord, R: Semigroup> MergeSorter<D, T, R> {
+    fn extract_into(
+        &mut self,
+        upper: AntichainRef<T>,
+        frontier: &mut Antichain<T>,
+    ) -> Vec<Vec<(D, T, R)>> {
+        // Flush pending data
+        crate::consolidation::consolidate_updates(&mut self.pending);
+        self.flush_pending();
+
+        let mut keep = self.empty();
+        let mut keep_frontier = Antichain::new();
+        let mut ship = self.empty();
+        let mut ship_list = Vec::default();
+
+        self.account(self.queue.iter().flatten(), -1);
+
+        // Walk all chains, separate ready data from data to keep.
+        for mut chain in std::mem::take(&mut self.queue).drain(..) {
+            let mut block_list = Vec::default();
+            let mut keep_list = Vec::default();
+            for (block_frontier, mut block) in chain.drain(..) {
+                // Is any data ready to be shipped?
+                let any = !PartialOrder::less_equal(&upper, &block_frontier.borrow());
+                // Is all data ready to be shipped?
+                let all = any && block.iter().all(|(_, t, _)| !upper.less_equal(t));
+
+                if all {
+                    // All data is ready, push what we accumulated, stash whole block.
+                    if !ship.is_empty() {
+                        block_list.push((Antichain::new(), std::mem::replace(&mut ship, self.empty())));
+                    }
+                    block_list.push((Antichain::new(), block));
+                } else if any {
+                    // Iterate block, sorting items into ship and keep
+                    for datum in block.drain(..) {
+                        if upper.less_equal(&datum.1) {
+                            frontier.insert_ref(&datum.1);
+                            keep_frontier.insert_ref(&datum.1);
+                            keep.push(datum);
+                            if keep.capacity() == keep.len() {
+                                // remember keep
+                                keep_list.push((std::mem::take(&mut keep_frontier), std::mem::replace(&mut keep, self.empty())));
+                            }
+                        } else {
+                            ship.push(datum);
+                            if ship.capacity() == ship.len() {
+                                // Ship is full, push in on the block list, get an empty one.
+                                block_list.push((Antichain::new(), std::mem::replace(&mut ship, self.empty())));
+                            }
+                        }
+                    }
+                    // Recycle leftovers
+                    self.recycle(block);
+                } else {
+                    // Keep the entire block.
+                    if !keep.is_empty() {
+                        keep_list.push((std::mem::take(&mut keep_frontier), std::mem::replace(&mut keep, self.empty())));
+                    }
+                    keep_list.push((block_frontier, block));
+                }
+            }
```

Comment:
This is potentially a good moment, as we go, to perform the "adjacent blocks" compaction. That would allow us to return memory to `self.empty()` eagerly, and have it available as we loop. Waiting until `maintain()` is not wrong, but it seems like it may cause memory to spike during `extract_into` and return down only after `maintain()`.